src/lightning/pytorch/utilities/data/dataset.py

```diff
@@ -0,0 +1,125 @@
+import os
+import tempfile
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional, Tuple
+
+from torch.utils.data import Dataset as TorchDataset
+
+from lightning.pytorch.utilities.data.fileio import OpenCloudFileObj
+from lightning.pytorch.utilities.data.get_index import get_index
+
+
+def get_aws_credentials():
+    """Gets AWS credentials from the current IAM role.
+
+    Returns:
+        credentials object to be used for file reading
+    """
+    from botocore.credentials import InstanceMetadataProvider
+    from botocore.utils import InstanceMetadataFetcher
+
+    provider = InstanceMetadataProvider(iam_role_fetcher=InstanceMetadataFetcher(timeout=1000, num_attempts=2))
+
+    credentials = provider.load()
+
+    os.environ["AWS_ACCESS_KEY"] = credentials.access_key
+    os.environ["AWS_SECRET_KEY"] = credentials.secret_key
+    os.environ["AWS_SESSION_TOKEN"] = credentials.token
+
+    return credentials
+
+
+class LightningDataset(TorchDataset):
+    """Dataset wrapper for optimized dataloading.
```

Comment:
can we pls add some more detail docs with use cases as this is going to be public API